backend/src/main/kotlin/code/nebula/cipherquest/service/GameService.kt

```diff
@@ -21,53 +22,65 @@ class GameService(
         }
-    }
 
     /**
      * Check if the user has spent all their coins, and return the game over message if so.
      */
-    private fun gameOver(userToQuery: Pair<UserLevel, String>): BotAnswer? = if (userToQuery.first.coins <= 0) BotAnswer
-        .buildGameOverMessage(userToQuery.first) else null
+    private fun gameOver(userToQuery: Pair<UserLevel, String>): BotAnswer? =
+        if (userToQuery.first.coins <= 0) {
+            BotAnswer
+                .buildGameOverMessage(userToQuery.first)
+        } else {
+            null
+        }
 
     /**
      * Go ahead to the next turn in the game. Pass the user's message to the chat client and return the response.
      */
-    private fun gameNextTurn(userToQuery: Pair<UserLevel, String>): String = chatClient
-        .prompt()
-        .system { sp -> sp.param("userId", userToQuery.first.userId).param("level", userToQuery.first.level) }
-        .user(userToQuery.second)
-        .advisors { a ->
-            a
-                .param(CHAT_MEMORY_CONVERSATION_ID_KEY, userToQuery.first.userId)
-                .param(CHAT_MEMORY_RETRIEVE_SIZE_KEY, 20)
-                .param(
-                    QuestionAnswerAdvisor.FILTER_EXPRESSION,
-                    "type == '${DocumentType.DOCUMENT}' && level <= " +
-                        "${userToQuery.first.level}",
-                )
-        }
-        .call()
-        .content()
+    private fun gameNextTurn(userToQuery: Pair<UserLevel, String>): String =
+        chatClient
+            .prompt()
+            .system { sp -> sp.param("userId", userToQuery.first.userId).param("level", userToQuery.first.level) }
+            .user(userToQuery.second)
+            .advisors { a ->
+                a
+                    .param(CHAT_MEMORY_CONVERSATION_ID_KEY, userToQuery.first.userId)
+                    .param(CHAT_MEMORY_RETRIEVE_SIZE_KEY, 20)
+                    .param(
+                        QuestionAnswerAdvisor.FILTER_EXPRESSION,
+                        "type == '${DocumentType.DOCUMENT}' && level <= " +
+                            "${userToQuery.first.level}",
+                    )
+            }.call()
+            .content()
 
-    fun play(userId: String, userMessage: String): BotAnswer {
+    fun play(
+        userId: String,
+        userMessage: String,
+    ): BotAnswer {
         val userLevel = userLevelService.getLevelByUser(userId)
 
         return listOf(::gameWon, ::gameOver, ::gameWin).firstNotNullOfOrNull { fn -> fn(Pair(userLevel, userMessage)) }
             ?: gameNextTurn(Pair(userLevel, userMessage)).let { response ->
-            val user = userLevelService.decreaseCoins(userId)
-            return BotAnswer.build(response, user)
-        }
+                val user = userLevelService.decreaseCoins(userId)
+                return BotAnswer.build(response, user)
+            }
     }
+
+    fun calculateScoreboard(): List<ScoreboardEntry> = userLevelService.calculateScoreboard()
```

Comment:
Is it really necessary to go through this service to calc the score?

Reply:
It is not necessary, but it is also not a problem. It could be used in the future to apply more business logic to the functionality